### PR TITLE
[WIP] add rc transfer statistics

### PR DESCRIPTION
**Pull request by seuffert**
implements #2280
 
added `rclone rc stats/transfers `and `rclone rc stats/reset` 

Is there more information we could give for the files in "_checking_" ?

@ncw during testing i noticed that sometimes a _transferring_ file vanishes from the statistics seconds after it was started. Any idea why? The stream continues without error and the global byte counter steadily increases.

### CODE REVIEW

File: fs/accounting/stats.go [hunk 1]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
+	g["errors"] = s.errors
+	g["deletes"] = s.deletes
+	s.mu.RUnlock()
+
+	r := make(map[string]map[string]interface{})
+	s.transferring.mu.RLock()
+	for name := range s.transferring.items {
+		acc := s.inProgress.get(name)
+		r[name] = make(map[string]interface{})
+		r[name]["inProgress"] = (acc != nil)
```
Comment (ncw) [inline fs/accounting/stats.go]: Should probably use snake case `in_progress` for consistency here

File: fs/accounting/stats.go [hunk 2]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
+	out["status"] = "ok"
+	out["global"] = g
+	out["transferring"] = t
+	out["checking"] = c
+	return out, nil
+}
+
+// Map convert the StatsInfo to a map for rc stats
+func (s *StatsInfo) Map() (map[string]interface{}, map[string]map[string]interface{}, map[string]map[string]interface{}, error) {
+	dataRateMultiplier := 1.0
```
Comment (ncw) [inline fs/accounting/stats.go]: I think I'd lose the `dataRateMultipler` and just define the output to be in bytes/s

File: fs/accounting/stats.go [hunk 3]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
+			r[name]["speed_moving_avg"] = speedMavg * dataRateMultiplier
+			r[name]["speed"] = speedBps * dataRateMultiplier
+			if etaOk {
+				r[name]["eta"] = eta
+			}
+			percentageDone := 0.0
+			if bytes > 0 {
+				percentageDone = 100 * float64(bytes) / float64(size)
+			}
+			r[name]["percentageDone"] = percentageDone
```
Comment (ncw) [inline fs/accounting/stats.go]: snake case here too

File: fs/accounting/stats.go [hunk 4]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
+		return out, errors.Errorf("error while getting bw stats")
+	}
+	out["status"] = "ok"
+	out["global"] = g
+	out["transferring"] = t
+	out["checking"] = c
+	return out, nil
+}
+
+// Map convert the StatsInfo to a map for rc stats
```
Comment (ncw) [inline fs/accounting/stats.go]: Can you write what it returns here too please - briefly describe the global, transferring and checking maps

File: fs/accounting/stats.go [hunk 5]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
+	})
+
+	return s
+}
+
+// reset statistics as requested by rc stats/reset
+func (s *StatsInfo) bwResetStats(in rc.Params) (out rc.Params, err error) {
+	out = make(rc.Params)
+
+	s.mu.Lock()
```
Comment (ncw) [inline fs/accounting/stats.go]: Note that we have StatsInfo.ResetCounters already...

File: fs/accounting/stats.go [hunk 6]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
+	s.bytes = 0
+	s.errors = 0
```
Comment (ncw) [inline fs/accounting/stats.go]: Note that resetting errors to 0 potentially changes the exit code which may/may not be desirable!  It is OK if we document it I guess..

File: fs/accounting/stats.go [hunk 7]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
+	s.checks = 0
+	s.transfers = 0
+	s.deletes = 0
+	s.start = time.Now()
```
Comment (ncw) [inline fs/accounting/stats.go]: Why do you want to reset the start time?

File: fs/accounting/stats.go [hunk 8]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
+		Help: `
+Resets the global transfer statistic counters.
+`,
+	})
+
+	return s
+}
+
+// reset statistics as requested by rc stats/reset
+func (s *StatsInfo) bwResetStats(in rc.Params) (out rc.Params, err error) {
```
Comment (ncw) [inline fs/accounting/stats.go]: This should probably be `rcResetStats`

File: fs/accounting/stats.go [hunk 9]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
+	s.start = time.Now()
+	s.mu.Unlock()
+
+	out["status"] = "ok"
+	out["message"] = "Global transfer statistic counters were reset to Zero"
+	return out, nil
+}
+
+// create map with statistics for rc stats/transfers
+func (s *StatsInfo) bwStats(in rc.Params) (out rc.Params, err error) {
```
Comment (ncw) [inline fs/accounting/stats.go]: This should probably be `rcStats` to fit in with the naming conventions elsewhere

File: fs/accounting/stats.go [hunk 10]
```diff
@@ -37,12 +39,130 @@ type StatsInfo struct {
...
 		start:        time.Now(),
 		inProgress:   newInProgress(),
 	}
+
+	rc.Add(rc.Call{
+		Path:  "stats/transfers",
+		Fn:    s.bwStats,
+		Title: "Get transfer stats",
+		Help: `
+Show statistics for all transfers.
```
Comment (ncw) [inline fs/accounting/stats.go]: More docs here would be nice :-)

**Comment by ncw** [comment]
Hi!  I just wondered if you had time to finish this off?  If not I can finish it off for you.

Thanks

Nick

**Comment by seuffert** [comment]
Hi Nick, vacation :) i will have a look at it in the next few days. if you need it fast just go ahead, i dont mind :) 
I was reading along in the forums and thought about adding some more stats , e.g. from pending uplads if possible

**Comment by ncw** [comment]
No hurry - happy to see what you come up with :-)

**Comment by ncw** [comment]
I'm going to close this now since the PR is now unmergeable and has been hanging around for too long :-(

If you'd like to continue, please open a new PR.

Apologies

**Comment by seuffert** [comment]
@ncw thanks. will do.

**Comment by ncw** [comment]
Thanks for your understanding and looking forward to see what you come up with!

**Pull request closed without merging**